src/applications/disability-benefits/526EZ/pages/primaryAddress.js

```diff
@@ -37,364 +45,30 @@ function validateMilitaryState(errors, state, formData, schema, messages, option
   const isMilitaryCity = MILITARY_CITIES.includes(
     _.get(formData, `veteran[${options.addressName}].city`, '').trim().toUpperCase()
   );
-  const isMilitaryState = MILITARY_STATES.includes(state);
+  const isMilitaryState = MILITARY_STATE_VALUES.includes(state);
   if (isMilitaryCity && !isMilitaryState) {
     errors.addError('State must be AA, AE, or AP when using a military city');
   }
 }
 
 const hasForwardingAddress = (formData) => (_.get(formData, 'veteran[view:hasForwardingAddress]', false));
 
-const states = [
-  'AL',
-  'AK',
-  'AS',
-  'AZ',
-  'AR',
-  'AA',
-  'AE',
-  'AP',
-  'CA',
-  'CO',
-  'CT',
-  'DE',
-  'DC',
-  'FM',
-  'FL',
-  'GA',
-  'GU',
-  'HI',
-  'ID',
-  'IL',
-  'IN',
-  'IA',
-  'KS',
-  'KY',
-  'LA',
-  'ME',
-  'MH',
-  'MD',
-  'MA',
-  'MI',
-  'MN',
-  'MS',
-  'MO',
-  'MT',
-  'NE',
-  'NV',
-  'NH',
-  'NJ',
-  'NM',
-  'NY',
-  'NC',
-  'ND',
-  'MP',
-  'OH',
-  'OK',
-  'OR',
-  'PW',
-  'PA',
-  'PR',
-  'RI',
-  'SC',
-  'SD',
-  'TN',
-  'TX',
-  'UT',
-  'VT',
-  'VI',
-  'VA',
-  'WA',
-  'WV',
-  'WI',
-  'WY',
-  'PI',
-  'UM'
-];
-
-const stateLabels = {
-  AL: 'Alabama',
-  AK: 'Alaska',
-  AS: 'American Samoa',
-  AZ: 'Arizona',
-  AR: 'Arkansas',
-  AA: 'Armed Forces Americas (AA)',
-  AE: 'Armed Forces Europe (AE)',
-  AP: 'Armed Forces Pacific (AP)',
-  CA: 'California',
-  CO: 'Colorado',
-  CT: 'Connecticut',
-  DE: 'Delaware',
-  DC: 'District Of Columbia',
-  FM: 'Federated States Of Micronesia',
-  FL: 'Florida',
-  GA: 'Georgia',
-  GU: 'Guam',
-  HI: 'Hawaii',
-  ID: 'Idaho',
-  IL: 'Illinois',
-  IN: 'Indiana',
-  IA: 'Iowa',
-  KS: 'Kansas',
-  KY: 'Kentucky',
-  LA: 'Louisiana',
-  ME: 'Maine',
-  MH: 'Marshall Islands',
-  MD: 'Maryland',
-  MA: 'Massachusetts',
-  MI: 'Michigan',
-  MN: 'Minnesota',
-  MS: 'Mississippi',
-  MO: 'Missouri',
-  MT: 'Montana',
-  NE: 'Nebraska',
-  NV: 'Nevada',
-  NH: 'New Hampshire',
-  NJ: 'New Jersey',
-  NM: 'New Mexico',
-  NY: 'New York',
-  NC: 'North Carolina',
-  ND: 'North Dakota',
-  MP: 'Northern Mariana Islands',
-  OH: 'Ohio',
-  OK: 'Oklahoma',
-  OR: 'Oregon',
-  PW: 'Palau',
-  PA: 'Pennsylvania',
-  PR: 'Puerto Rico',
-  RI: 'Rhode Island',
-  SC: 'South Carolina',
-  SD: 'South Dakota',
-  TN: 'Tennessee',
-  TX: 'Texas',
-  UT: 'Utah',
-  VT: 'Vermont',
-  VI: 'Virgin Islands',
-  VA: 'Virginia',
-  WA: 'Washington',
-  WV: 'West Virginia',
-  WI: 'Wisconsin',
-  WY: 'Wyoming',
-  PI: 'Rizal state',
-  UM: 'United States Minor Outlying Islands'
-};
+const addressUISchema = (addressName, title) => {
+  const updateStates = (form) => {
+    const currentCity = _.get(form, `veteran.${addressName}.city`, '').trim().toUpperCase();
+    if (MILITARY_CITIES.includes(currentCity)) {
```

Comment:
Should we check that a state value doesn't already exist? So if there's no state value, restrict the enums, but if there's already a state value, simply provide a validation message upon submit instead?

Reply:
Good question. So here are the things that happen:
- When no state is present and the user enters a military city code:
    1. The enum list is restricted
    2. The 'empty' state option remains selected
    3. No validation errors fire yet, because the state field hasn't been touched
- When state is a non-military state and the user enters a military city code
    1. The enum list is restricted
    2. The selected state option is cleared and set to the empty option
    3. No validation errors fire yet, because the state field hasn't been touched
    4. Once the state field has been touched and the veteran has tabbed over it (i.e., kept the 'empty' selection and kept going to the other fields in the form), a validation message is shown
- When state is a non-military state, the city is changed to a military city code, and the state field has already been touched
    1. The enum list is restricted
    2. The selected state option is cleared and set to the empty option
    3. The validation error fires immediately, prompting the user to select a valid state now that they've changed the city to a military city code

I think this behavior is actually what I'd want, as a user. Don't bug me before I've had a chance to fill out the field, but let me know that I've done something invalid once I've selected an option that invalidates a previous selection in another field.

Reply:
Thanks for listing it out @markgreenburg, I'm mostly curious about clearing user input, since I'm not familiar with us doing that elsewhere for situations where the field is the same, but that doesn't mean it's necessarily something we have to avoid. As a user I might be slightly confused why a field I thought was filled is suddenly blank and has a validation error. Maybe just providing the validation error, once the military city is selected could be less jarring in that case?

Reply:
I dunno; the behavior in the existing approach is basically what comes with Forms out of the box. The implementation matches the [Forms cookbook example](https://github.com/department-of-veterans-affairs/vets-website/blob/master/docs/schemaform/form-config.md#i-want-to-change-the-options-of-a-dropdown-based-on-some-other-field-data). Not sure if you've tried to run this branch locally, but if not I'd be curious to hear if you still think the behavior is not intuitive after giving it a spin.
I see where you're coming from, though - one thing to keep in mind is that the city input comes before the state input, so in typical usage, the thing you just entered wouldn't disappear from the form just because the next thing you entered invalidates it.

For what it's worth, the Profile address widget and Letters app both clear non-military states once you enter a military city as well. Now that I think about it, I'm not certain how you'd restrict the options dropdown without clearing the existing option - I think you'd have to add it to the list dynamically, which could be strange (e.g., the list could look like `['Armed Forces Americas', ... , 'Michigan']`).